In [1]:
import pandas as pd
import numpy as np

In [6]:
df = pd.read_csv("data/train.csv")

In [7]:
# Making all spaces and black characters as NaN
df = df.replace(r'^\s*$', np.nan, regex=True)

In [8]:
#Filling with mode for the other columns
df['STATE'].fillna(df['STATE'].mode()[0], inplace=True)
df['OCCUPATION'].fillna(df['OCCUPATION'].mode()[0], inplace=True)
df['INCOME_GROUP'].fillna(df['INCOME_GROUP'].mode()[0], inplace=True)
df['LOYALTY_PROGRAM'].fillna(df['LOYALTY_PROGRAM'].mode()[0], inplace=True)

In [9]:
df

ID           STATE  AGE     OCCUPATION INCOME_GROUP  \
0            1   Uttar Pradesh   28  Self employed          Low   
1            2  Madhya Pradesh   27       Salaried          Low   
2            3     West Bengal   62  Self employed         High   
3            4      Tamil Nadu   27       Salaried          Low   
4            5   Uttar Pradesh   40  Self employed         High   
...        ...             ...  ...            ...          ...   
149995  149996           Assam   29       Salaried          Low   
149996  149997  Andhra Pradesh   53  Self employed       Medium   
149997  149998          Odisha   31  Self employed          Low   
149998  149999  Andhra Pradesh   54  Self employed         High   
149999  150000     Maharashtra   60  Self employed       Medium   

        CUSTOMER_SINCE LOYALTY_PROGRAM  PAST_PURCHASE PURCHASE  
0                 2018              No           4937       No  
1                 2018              No          10834       No  
2                 2013              No           8610      Yes  
3                 2018              No          14374      Yes  
4                 2015              No          13158       No  
...                ...             ...            ...      ...  
149995            2017              No           6241       No  
149996            2010              No          29447       No  
149997            2018              No           4856       No  
149998            2010             Yes          10031      Yes  
149999            2013              No           7761      Yes  

[150000 rows x 9 columns]

In [10]:
#Dropping the ID column 
df = df.drop(['ID'], axis = 1)

In [11]:
df['LOYALTY_PROGRAM'] = df['LOYALTY_PROGRAM'].map({'Yes': 1, 'No': 0})
df['PURCHASE'] = df['PURCHASE'].map({'Yes': 1, 'No': 0})

In [12]:
from sklearn.model_selection import train_test_split

In [13]:
#Split the dataset into 70(train) - 30(test)
X = df.drop(['PURCHASE'], axis=1)
y = df['PURCHASE']

In [14]:
import category_encoders as ce

In [15]:
# Define catboost encoder
cbe_encoder = ce.cat_boost.CatBoostEncoder()
  
# Fit encoder and transform the features
cbe_encoder.fit(X, y)
train_cbe = cbe_encoder.transform(X)

In [16]:
X_train, X_test, y_train, y_test = train_test_split(train_cbe, y, test_size = 0.3, random_state = 0)

In [18]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score
from xgboost import XGBClassifier
# fit model no training data
model = XGBClassifier()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
predictions = [round(value) for value in y_pred]
# evaluate predictions
accuracy = accuracy_score(y_test, predictions)
print("Accuracy: %.2f%%" % (accuracy * 100.0))
print("F1 Score: ",f1_score(y_test, predictions))

c:\users\raksha s\appdata\local\programs\python\python39\lib\site-packages\xgboost\sklearn.py:1146: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[01:14:09] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.4.0/src/learner.cc:1095: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
Accuracy: 78.42%
F1 Score:  0.29994952772370037


In [19]:
from catboost import CatBoostClassifier
# fit model no training data
model = CatBoostClassifier()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
predictions = [round(value) for value in y_pred]
# evaluate predictions
accuracy = accuracy_score(y_test, predictions)
print("Accuracy: %.2f%%" % (accuracy * 100.0))
print("F1 Score: ",f1_score(y_test, predictions))

Learning rate set to 0.075159
0:	learn: 0.6518917	total: 186ms	remaining: 3m 5s
1:	learn: 0.6177848	total: 226ms	remaining: 1m 52s
2:	learn: 0.5900159	total: 258ms	remaining: 1m 25s
3:	learn: 0.5691576	total: 287ms	remaining: 1m 11s
4:	learn: 0.5512466	total: 315ms	remaining: 1m 2s
5:	learn: 0.5352269	total: 345ms	remaining: 57.1s
6:	learn: 0.5226845	total: 374ms	remaining: 53s
7:	learn: 0.5120381	total: 402ms	remaining: 49.9s
8:	learn: 0.5035280	total: 430ms	remaining: 47.4s
9:	learn: 0.4958808	total: 459ms	remaining: 45.4s
10:	learn: 0.4900462	total: 488ms	remaining: 43.9s
11:	learn: 0.4848543	total: 514ms	remaining: 42.3s
12:	learn: 0.4809233	total: 542ms	remaining: 41.2s
13:	learn: 0.4770527	total: 572ms	remaining: 40.3s
14:	learn: 0.4745073	total: 599ms	remaining: 39.3s
15:	learn: 0.4716452	total: 627ms	remaining: 38.6s
16:	learn: 0.4694284	total: 657ms	remaining: 38s
17:	learn: 0.4668116	total: 686ms	remaining: 37.4s
18:	learn: 0.4652332	total: 716ms	remaining: 37s
19:	learn: 0.4

In [20]:
print("Accuracy: %.2f%%" % (accuracy * 100.0))
print("F1 Score: ",f1_score(y_test, predictions))

Accuracy: 78.64%
F1 Score:  0.30704915669597804


In [21]:
from lightgbm import LGBMClassifier
# fit model no training data
model = LGBMClassifier()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
predictions = [round(value) for value in y_pred]
# evaluate predictions
accuracy = accuracy_score(y_test, predictions)
print("Accuracy: %.2f%%" % (accuracy * 100.0))
print("F1 Score: ",f1_score(y_test, predictions))

Accuracy: 78.73%
F1 Score:  0.2717383035374667


We see that when we make use of all the columns and impute the values using the mode, we get an F1 score of 0.30. This is still not a very good score. We need to look for other methods where we can improve this F1 score.  